### ISSUE TITLE
Standalone DSN parser

### ISSUE THREAD

**Issue by derrabus** [doctrine/dbal#5814]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no

#### Summary

In #5697, we have deprecated driver aliases for database URLs. One use-case of those aliases was to be able to parse a credentials URL like `mysql://my_user:my_password@my.database-host.local` that are usually provided by a PaaS services. With the changes on 4.0, those database URLs would become specific to Doctrine because we would have to use `mysqli://` or `pdo-mysql://` as scheme.

My idea would be to deprecate the `url` option completely and replace it with a standalone DSN parser that translates a given DSN into an array of options that can be passed to `DriverManager`. This parser can be configured with a scheme-to-driver mapping (e.g. `['mysql' => 'pdo_mysql']`).

I see several advantages of this approach:

* The DBAL does not decide which driver to pick for a database engine, if multiple drivers exist.
* The array can be modified before it's passed to DriverManager.
  * Adding settings that the DSN does not cover is simple and discoverable.
  * It is easier to see how settings from the DSN can be overridden.
* The array can be cached by the application which means we don't need to parse the DSN on each request.
* `DriverManager`'s internal logic gets a little bit simpler.

**Comment by derrabus** [comment]
@stof, I think your input can help on this issue.

**Comment by stof** [comment]
> * The array can be cached by the application which means we don't need to parse the DSN on each request.

this is exactly the opposite of the needs of Symfony though. The whole point is to avoid caching the parsed credentials so that you don't need to clear the cache when credentials change (on most PaaS, clearing cache requires a new deployment, while credentials are injected as env variable in the container without rebuilding the image when they change).

However, having it as a separate class that parses the URL and merges that with an array of options (with the explicit options winning over the URL, at least when non-null) might solve it

**Comment by derrabus** [comment]
I wrote "_can_ be cached". Of course, you shouldn't do it if the DSN can change at any time without cache invalidation. Caching those values would be the responsibility of the application that configures the DBAL connection.

The idea is that the application contains bootstrapping logic like this:

```php
$parameters = $dsnParser->parse($dsn);

// Post-process the parameters as you see fit!

$connection = DriverManager::getConnection($parameters);
```

In case of Symfony, that logic would reside in DoctrineBundle.

**Comment by derrabus** [comment]
I've started working on this feature in #5843.

**Comment by derrabus** [comment]
Fixed via #5843
